src/SDK/Trace/SpanProcessor/BatchSpanProcessor.php

```diff
@@ -4,142 +4,183 @@
-            );
+        if ($maxQueueSize <= 0) {
+            throw new InvalidArgumentException(sprintf('Maximum queue size (%d) must be greater than zero', $maxQueueSize));
+        }
+        if ($scheduledDelayMillis <= 0) {
+            throw new InvalidArgumentException(sprintf('Scheduled delay (%d) must be greater than zero', $scheduledDelayMillis));
+        }
+        if ($exportTimeoutMillis <= 0) {
+            throw new InvalidArgumentException(sprintf('Export timeout (%d) must be greater than zero', $exportTimeoutMillis));
+        }
+        if ($maxExportBatchSize <= 0) {
+            throw new InvalidArgumentException(sprintf('Maximum export batch size (%d) must be greater than zero', $maxExportBatchSize));
         }
+        if ($maxExportBatchSize > $maxQueueSize) {
+            throw new InvalidArgumentException(sprintf('Maximum export batch size (%d) must be less than or equal to maximum queue size (%d)', $maxExportBatchSize, $maxQueueSize));
+        }
+
+        $this->exporter = $exporter;
+        $this->clock = $clock;
+        $this->maxQueueSize = intdiv($maxQueueSize, $maxExportBatchSize);
+        $this->maxBatchSize = $maxQueueSize % $maxExportBatchSize;
+        $this->scheduledDelayNanos = $scheduledDelayMillis * 1_000_000;
+        $this->maxExportBatchSize = $maxExportBatchSize;
+        $this->autoFlush = $autoFlush;
+
+        $this->queue = new SplQueue();
+        $this->flush = new SplQueue();
     }
 
-    /**
-     * @inheritDoc
-     */
     public function onStart(ReadWriteSpanInterface $span, Context $parentContext): void
     {
     }
 
-    /**
-     * @inheritDoc
-     */
     public function onEnd(ReadableSpanInterface $span): void
     {
-        if (null === $this->exporter) {
+        if ($this->closed) {
             return;
         }
-
-        if (!$this->running) {
+        if (!$span->getContext()->isSampled()) {
             return;
         }
 
-        if ($span->getContext()->isSampled() && !$this->queueReachedLimit()) {
-            $this->queue[] = $span->toSpanData();
+        if (count($this->queue) === $this->maxQueueSize && count($this->batch) === $this->maxBatchSize) {
+            return;
         }
 
-        if ($this->bufferReachedExportLimit() || $this->enoughTimeHasPassed()) {
-            $this->forceFlush();
+        $this->batch[] = $span->toSpanData();
+        $this->nextScheduledRun ??= $this->clock->now() + $this->scheduledDelayNanos;
+
+        if (count($this->batch) === $this->maxExportBatchSize) {
+            $this->enqueueBatch();
+        }
+        if ($this->autoFlush && !$this->running && $this->shouldFlush()) {
+            $this->flush();
         }
     }
 
-    /** @inheritDoc */
     public function forceFlush(?CancellationInterface $cancellation = null): bool
     {
-        if (!$this->running || $this->exporter === null) {
-            return true;
+        if ($this->closed) {
+            return $this->flush->isEmpty();
```

Comment:
Not too sure about this, should either return `false` or call `::flush()` to allow recovering from a shutdown that was terminated by an exception.

Reply:
it makes sense that it would try to call `flush` again, the method is _force_ flush

Reply:
After rewriting the async implementation to match this PR: I think that preventing termination of the worker by swallowing/logging `::export()` exceptions (and `::forceFlush()` / `::shutdown()` exceptions if we cannot rethrow them into the original caller) makes the most sense. This would make recovery post-shutdown obsolete.